src/osd/ReplicatedBackend.cc

```diff
@@ -1665,24 +1693,44 @@ void ReplicatedBackend::submit_push_data(
   map<string, bufferlist> &omap_entries,
   ObjectStore::Transaction *t)
 {
-  hobject_t target_oid;
-  if (first && complete) {
-    target_oid = recovery_info.soid;
-  } else {
-    target_oid = get_parent()->get_temp_recovery_object(recovery_info.version,
+  hobject_t target_oid = get_parent()->get_temp_recovery_object(recovery_info.version,
```

Comment:
yeah，I preserve this at the first version.  It is really useful optimization, but we need to reconsider the truncate operation carefully. 
One case I can imagine is that
Assume object 4M A is mapping to osd.[1,2,3]
Osd.2 down first and truncate on osd.1 and osd.3 to 1M
Then new write overwrite at 3M~4M and still leave sparse between 1M~3M
When osd.2 comes up again,a directly overwrite will lead to osd.2 no sparse with old content between 1M~3M.  
One solution to deal with it is to log truncate Ops in pg_log distinguished with Modify Ops. But this will alter the logic of pg_log_entry_t a lots.
Any better suggestion and guides for this?  @athanatos